cmd/kubeadm/app/util/config/config.go

```diff
@@ -14,58 +14,54 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package cmd
+package util
 
 import (
 	"fmt"
+	"io/ioutil"
 	"net"
 
+	"k8s.io/apimachinery/pkg/runtime"
 	netutil "k8s.io/apimachinery/pkg/util/net"
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
 	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
 	tokenutil "k8s.io/kubernetes/cmd/kubeadm/app/util/token"
+	"k8s.io/kubernetes/pkg/api"
 
 	"github.com/blang/semver"
 )
 
 var minK8sVersion = semver.MustParse(kubeadmconstants.MinimumControlPlaneVersion)
 
-func setInitDynamicDefaults(cfg *kubeadmapi.MasterConfiguration) error {
+// SetInitDynamicDefaults set defaults that the API group defaulting can't (by fetching information from the internet, looking up network interfaces, etc.)
+func SetInitDynamicDefaults(cfg *kubeadmapi.MasterConfiguration) error {
 
-	// Choose the right address for the API Server to advertise. If the advertise address is localhost or 0.0.0.0, the default interface's IP address is used
+	// Choose the right address for the API Server to advertise. If the advertise address is localhost or 0.0.0.0 or empty, the default interface's IP address is used
 	// This is the same logic as the API Server uses
 	ip, err := netutil.ChooseBindAddress(net.ParseIP(cfg.API.AdvertiseAddress))
 	if err != nil {
 		return err
 	}
 	cfg.API.AdvertiseAddress = ip.String()
 
-	// Validate version argument
+	// Choose the kubernetes version. If empty fetch version information from release servers.
 	ver, err := kubeadmutil.KubernetesReleaseVersion(cfg.KubernetesVersion)
 	if err != nil {
 		return err
 	}
 	cfg.KubernetesVersion = ver
 
-	// Omit the "v" in the beginning, otherwise semver will fail
-	k8sVersion, err := semver.Parse(cfg.KubernetesVersion[1:])
+	// Validate version argument versus kubeadmconstants.MinimumControlPlaneVersion
+	k8sVersion, err := semver.Parse(cfg.KubernetesVersion[1:]) // Omit the "v" in the beginning, otherwise semver will fail
```

Comment:
unfortunately kubernetes imports v3.0.1 of `semver` and this version does not contains `ParseTolerant`method...